src/common/post.ts

```diff
@@ -446,9 +543,136 @@ export interface CreatePollPostProps
+  sourceIds: string[];
+}
+
+const MAX_MULTIPLE_POST_SOURCE_LIMIT = 4;
+
 const MAX_TITLE_LENGTH = 250;
 const MAX_CONTENT_LENGTH = 10_000;
 
+export const postInMultipleSourcesArgsSchema = z
+  .object({
+    title: z.string().max(MAX_TITLE_LENGTH).optional(),
+    content: z.string().max(MAX_CONTENT_LENGTH).optional(),
+    image: z.custom<Promise<FileUpload>>(),
+    sourceIds: z.array(z.string()).min(1).max(MAX_MULTIPLE_POST_SOURCE_LIMIT),
+    sharedPostId: z.string().optional(),
+  })
+  .extend(
+    pollCreationSchema
+      .pick({
+        options: true,
+        duration: true,
+      })
+      .partial().shape,
+  );
+
+export const getMultipleSourcesPostType = (
+  args: z.infer<typeof postInMultipleSourcesArgsSchema>,
+): PostType => {
+  if (args.options?.length) {
+    return PostType.Poll;
+  }
+
+  if (args.sharedPostId) {
+    return PostType.Share;
+  }
+
+  return PostType.Freeform;
+};
+
+export const checkIfUserPostInSourceDirectlyOrThrow = async (
+  con: DataSource,
+  { sourceId, userId }: Record<'sourceId' | 'userId', string>,
+) => {
+  const isSameUserSource = sourceId === userId;
+
+  if (isSameUserSource) {
+    return true;
+  }
+
+  const [source, squadMember] = await Promise.all([
+    con.getRepository(Source).findOneBy({
+      id: sourceId,
+    }),
+    con.getRepository(SourceMember).findOneBy({
+      userId,
+      sourceId,
+    }),
+  ]);
+
+  if (!source || !squadMember || source?.type !== SourceType.Squad) {
+    throw new ForbiddenError('Access denied!');
+  }
+
+  return canPostToSquad(source as SquadSource, squadMember);
+};
+
+export const createPostIntoSourceId = async (
+  ctx: AuthContext,
+  sourceId: string,
+  args: z.infer<typeof postInMultipleSourcesArgsSchema>,
+  entityManager?: EntityManager,
+): Promise<Pick<Post, 'id'>> => {
+  const type = getMultipleSourcesPostType(args);
+  const con = entityManager || ctx.con;
+  switch (type) {
+    case PostType.Share: {
+      await ctx.con
+        .getRepository(Post)
+        .findOneByOrFail({ id: args.sharedPostId });
+      return await createSharePost({
+        con,
+        ctx,
+        args: {
+          authorId: ctx.userId,
+          sourceId,
+          postId: args.sharedPostId!,
+          commentary: args.title,
+        },
+      });
+    }
+    case PostType.Poll: {
+      const id = await generateShortId();
+      const { options, ...pollArgs } = args;
+      return await createPollPost({
+        con,
+        ctx,
+        args: {
+          ...pollArgs,
+          id,
+          sourceId,
+          title: `${args.title}`,
+          authorId: ctx.userId,
+          pollOptions: options!.map((option) =>
+            ctx.con.getRepository(PollOption).create({
+              text: option.text,
+              numVotes: 0,
+              order: option.order,
+              postId: id!,
+            }),
+          ),
+        },
+      });
+    }
+    default: {
+      return await createFreeformPost(
```

Comment:
should this default to freeform, I think safer if its case freeform, and default is throw because its something we don't expect?

Reply:
Updated ✔️